### PR TITLE
Adds a check on pull requests for baking time

### PR DESCRIPTION
**Pull request by peternied**
### Description
After pull requests are updated, a 2 day waiting period is required to ensure plenty of time for additional reviews and comments.

This repository handles changes that are sensitive to the OpenSearch-Project involving many maintainers/contributors by using a tool to enforce the 'bake time' it removes maintainers from having to remember how long to wait and communicates to contributors when a pull request is ready to merge.

The underlying github actions for this change are in https://github.com/peternied/bake-time to make this functionality reusable in other repositories.

### Issues
- Related https://github.com/opensearch-project/.github/issues/149 

#### Problem
This .github repository represents OpenSearch-Project's policies, ideals, and intentions. This repository has no quality gates other than 2 approvals and a DCO check - making the way pull requests are accepted subjective based on which maintainers/contributors first responded.

#### Expectation
There should be mechanisms to ensure a minimum quality of the changes that are accepted into this repository.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: .github/workflows/start_bake_time.yml
Comment (peternied) [inline .github/workflows/start_bake_time.yml:12]: @opensearch-project/engineering-effectiveness I'm getting an error that the GITHUB_TOKEN doesn't have access to make these changes, do you have thoughts or a recommendations for how to resolve this?

```
    data: {
      message: 'Resource not accessible by integration',
      documentation_url: 'https://docs.github.com/rest/reference/issues#add-labels-to-an-issue'
    }
```

**Comment by peternied** [comment]
While troubleshooting the permissions issue, I'm would appreciate any feedback on the concept.

**Comment by dblock** [comment]
I really like this idea!

From the implementation POV I think you should combine start and finish actions into a single pass GHA that is smart enough to know what it's running against (PR/cron). As a user, I'd like to write a single workflow. 

```
name: Bake Time

on:
  workflow_dispatch:
  pull_request:
    types:
      - opened
      - synchronize
  schedule:
    - cron: '0 */1 * * *' # finish/start baking every hour

jobs:
  bake:
    runs-on: ubuntu-latest
    steps:
      - uses: peternied/bake-time/bake@v1
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
          bake-time: 48
```

File: .github/workflows/start_bake_time.yml
Comment (peternied) [reply .github/workflows/start_bake_time.yml:12]: It is strange as the there is another workflow that adds labels to issues - no problems https://github.com/opensearch-project/.github/blob/main/.github/workflows/add-untriaged.yml

I'll try more experimentation within this pull request - maybe I am passing the wrong token or maybe action/github-script works differently as a composite action

Comment (peternied) [reply .github/workflows/start_bake_time.yml]: Circling back on this, when a pull request triggers a github action it cannot perform _any_ write operations.  This is a good security measure, and in order to make this functional I had to drop all write actions during this trigger.  Now the check 'fails' on the pull request with the message saying how long it will take.

If we want a cleaner experience a bot will need to execute the actions.

**Comment by dblock** [comment]
This looks great! Is there any way to change the GHA to report pending status rather than a failing status? 

Reading https://dev.to/cardinalby/scheduling-delayed-github-action-12a6 it seems like you should be able to change the cron into a trigger that runs once, then dispatches itself/another job with a wait timer (https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#wait-timer) and repeats it until it runs out of time.

**Comment by peternied** [comment]
> Is there any way to change the GHA to report pending status rather than a failing status?

When using the pull request trigger via GitHub Actions, no.  Yes - by using a GitHub bot/webhook.

This is a trade-off as this version will works with the fewest permissions being in play during a pull request and visibility in the GitHub action script/logs.

> And even say how much time is left to wait / when the last check ran?
Nice idea, the scheduled action could update a hours remaining value and make it displaying pending instead of failed

Representing the 'last ran time' is difficult with static text, it would need to be in UTC otherwise we run into timezone problems, otherwise it would say 'last run an hour ago' or something less useful

> cron into a trigger that runs once
Nice find, we would need a personal access token to perform the workflow dispatch, looks like it would only invoke the work once when the bake time was over - possible if we built up a bot/webhook.

**Comment by peternied** [comment]
>> Is there any way to change the GHA to report pending status rather than a failing status?

> Nice idea, the scheduled action could update a hours remaining value and make it displaying pending instead of failed

Following up on this, the check cannot have its status modified from 'completed' -> 'pending'.  so the UX does not look any cleaner 😞 

> And even say how much time is left to wait / when the last check ran?

Done, example:
image


@dblock @CEHENKLE @mch2 @saratvemulapalli What are your thoughts on merging this change?

**Comment by peternied** [comment]
Great idea, I have published 'bake-time' in the marketplace - as always feel to file issues and/or make pull requests

https://github.com/marketplace/actions/bake-time

**Comment by saratvemulapalli** [comment]
I still didnt really understand the intent, maintainers of repo do this anyway right? 
Is this a mechanism to enforce it?

**Review by dblock — COMMENTED** [review]
Is it supposed to turn green after 48 hours in this PR? I'm ready to approve it otherwise ;)

**Comment by peternied** [comment]
This background job that marks the 'Bake time / Baking pull request...' complete. And it will not run until after this job is merged, so this PR will need to be merged even with a failing bake time check (Irony!).

**Review by dblock — APPROVED** [review]

**Comment by peternied** [comment]
> Should I override branch protection and merge this?

Please do!

**Pull request merged**
